client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java

```diff
@@ -197,14 +198,20 @@ private int computeReadLength(long currentFetchOffset) {
         return Math.toIntExact(Math.min(currentReadLength, numberOfBytesRemaining));
     }
 
+    @GuardedBy("$lock")
+    private void cancelOutstandingRequest() {
+        outstandingRequest.cancel(true);
```

Comment:
This is a complicated case:
If the buffer has completed but has not had any slices pulled out of it we need to call `releaese()` if it does, then those could become invalid. So we need to make sure they don't like longer than the scope of `$lock`